nixos/doc/manual/development/writing-nixos-tests.section.md

```diff
@@ -417,3 +478,183 @@ import ./make-test-python.nix
 ```
 
 In that case, `numpy` is chosen from the generic `python3Packages`.
+
+## Constructing a test matrix {#sec-nixos-test-matrix}
+
+You may want to run variations of your test, or perhaps split long tests into
+a few shorter, focused tests.
+
+While you could roll your own solution with `mapAttrs`, this is actually
+non-trivial because of the requirements imposed by `nix-build`, hydra and
+`passthru.tests`. So instead, we recommend to use the `matrix` test options that
+take care of this for you.
+
+These options work by duplicating the entire test module configuration below
+the `matrix.<decision>.choice.<choice>.module` options.
+Laziness makes this pattern efficient.
+
+The test framework ([`runTest`](#sec-calling-nixos-tests)) takes care of
+traversing the choices one by one, constructing a tree of attribute sets where
+each edge represents a choice.
+
+A parameterized test may look as follows:
+
+```nix
+{ lib, backend, ... }: {
+  name = "foo-${backend}";
+
+  matrix.backend.choice.smtp.module = {};
+  matrix.backend.choice.rest.module = {};
+
+  defaults.services.foo.backend = backend;
+
+  nodes = …;
+  testScript = ''
+    ${lib.optionalString (backend == "rest") ''
+      # extra setup for the rest backend
+    ''}
+    # …
+  '';
+}
+```
+
+`runTest` transforms this to:
+
+```nix
+{
+  backend-rest = «derivation /nix/store/…-foo-rest.drv»;
+  backend-smtp = «derivation /nix/store/…-foo-smtp.drv»;
+
+  # make `nix-build` traverse this attrset to evaluate nested tests
+  recurseForDerivations = true;
+  # …
+}
+```
+
+It has picked up on the `matrix` and only returns the derivations constructed for each choice.
```

Comment:
```suggestion
It has picked up on the `matrix` and only returns the derivations constructed for each choice.

The arguments to the test module, which is called by `runTest` to create each configuration in the matrix, are built up as follows:
`matrix.<decision>` corresponds to the argument name to the test module.
`choice.<choice>` corresponds to the named argument's value as a string.
```

Something along these lines would have made clear to me from the start how the attributes are related to the test module arguments.

Reply:
Changing to

```suggestion
It has picked up on the `matrix` and only returns the derivations constructed for each choice.

The arguments to the test module, which is called by `runTest` to create each configuration in the matrix, are built up as follows:
 - `<decision>` in `matrix.<decision>` corresponds to the argument name to the test module.
 - `<choice>` in `choice.<choice>` corresponds to the named argument's value as a string. Non-string values can be set with `_module.args` or other options inside [`choice.<choice>.module`](#opt-matrix._name_.choice._name_.module).

```